### PR TITLE
rpmsg: reserve 0-1023 for predefined services in rpmsg_create_ept

### PR DESCRIPTION
**Pull request by xiaoxiang781216**
and allocate the dynamic address from 1024 like Linux kernel

Signed-off-by: Xiang Xiao <xiaoxiang@xiaomi.com>

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
caller can still pass the fixed address from 0 to 1023, but rpmsg_create_ept allocate the dynamic address from 1024.

**Comment by arnopo** [comment]
I think here it is important that we are aligned first on the Linux implementation.
In [rpmsg/virtio_rpmsg_bus.c](https://elixir.bootlin.com/linux/latest/source/drivers/rpmsg/virtio_rpmsg_bus.c#L162) the reserved address are defined by:
```
/*
 * Local addresses are dynamically allocated on-demand.
 * We do not dynamically assign addresses from the low 1024 range,
 * in order to reserve that address range for predefined services.
 */
#define RPMSG_RESERVED_ADDRESSES	(1024)

```
So here the main point here is to understand  what is a predefined service

Now let's have look at [__rpmsg_create_ept](https://elixir.bootlin.com/linux/latest/source/drivers/rpmsg/virtio_rpmsg_bus.c#L232) to understand how RPMSG_RESERVED_ADDRESSES is used

```
	/* do we need to allocate a local address ? */
	if (addr == RPMSG_ADDR_ANY) {
		id_min = RPMSG_RESERVED_ADDRESSES;
		id_max = 0;
	} else {
		id_min = addr;
		id_max = addr + 1;
	}

```
Regarding this code a predefined service is a service with a fixed address. And as this function can be called through rpmsg_create_ept API, any user can define a "predefined" service by  specifying the  chinfo.src address.
So if i well understand, the Linux implementation rule is:
-  service based on a service name will have an address >= 1024
-  address from 0 to 1023 are reserved for services with a fixed address on endpoint creation. fixed address can be used either par a core or a user service.

Did i miss something?

**Comment by xiaoxiang781216** [comment]
Yes, your understand is correct

**Comment by arnopo** [comment]
This version seems to me a good compromise that would match with Linux implementation, while keeping compatibility with existing implementation:
- address 0  to 1023: predefined address only, no control on source address
- address 1024 to 1151 shared between dynamically allocated and predefined address with a control of the address
-  higher than 1151 not supported 

I just need to do few more test on my side before approving it.

Then the title of this patch would need to be reworked as rpmsg_create_ept  can allocated reserved address.
Some comments in code describing the address management would also be useful to understand the new behavior, especially the fact that address between 0 and 1023 are not checked.

@edmooring: could you review it and give your opinion please?

**Comment by xiaoxiang781216** [comment]
> Then the title of this patch would need to be reworked as rpmsg_create_ept can allocated reserved address.
> Some comments in code describing the address management would also be useful to understand the new behavior, especially the fact that address between 0 and 1023 are not checked.

@arnopo done.

**Comment by xiaoxiang781216** [comment]
@arnopo could you merge the follow patch first:
https://github.com/OpenAMP/open-amp/pull/209
https://github.com/OpenAMP/open-amp/pull/207
https://github.com/OpenAMP/open-amp/pull/204
Then I can rebase other patch on top of them.

File: lib/rpmsg/rpmsg.c [hunk 1]
```diff
@@ -29,7 +29,7 @@ static uint32_t rpmsg_get_address(unsigned long *bitmap, int size)
 
 	nextbit = metal_bitmap_next_clear_bit(bitmap, 0, size);
```
Comment (edmooring) [inline lib/rpmsg/rpmsg.c:30]: Maybe add a comment above that the bitmap is used to allocate addresses above RPMSG_RESERVED_ADDRESSES. It took me some time to see how it worked.

File: lib/include/openamp/rpmsg.h [hunk 2]
```diff
@@ -25,22 +25,23 @@ extern "C" {
...
-#define RPMSG_ERR_ADDR		(RPMSG_ERROR_BASE - 7)
+#define RPMSG_SUCCESS			0
+#define RPMSG_ERROR_BASE		-2000
+#define RPMSG_ERR_NO_MEM		(RPMSG_ERROR_BASE - 1)
+#define RPMSG_ERR_NO_BUFF		(RPMSG_ERROR_BASE - 2)
+#define RPMSG_ERR_PARAM			(RPMSG_ERROR_BASE - 3)
+#define RPMSG_ERR_DEV_STATE		(RPMSG_ERROR_BASE - 4)
+#define RPMSG_ERR_BUFF_SIZE		(RPMSG_ERROR_BASE - 5)
+#define RPMSG_ERR_INIT			(RPMSG_ERROR_BASE - 6)
+#define RPMSG_ERR_ADDR			(RPMSG_ERROR_BASE - 7)
```
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply lib/include/openamp/rpmsg.h:44]: Since the new macro(RPMSG_RESERVED_ADDRESSES) is too long.

**Pull request merged**